### ISSUE TITLE
[Question] Example of applying statechart yaml or statechart python for a Django model?

### ISSUE THREAD

**Issue by simkimsia** [AlexandreDecan/sismic#121]
I read https://github.com/AlexandreDecan/sismic/issues/9

In the same spirit, I wonder if there can be an example of applying statechart (either as yaml or python API) as a way to write BDD tests on a Django model?

I can contribute to the example but I need some guidance with regards to sismic.

**Comment by AlexandreDecan** [comment]
Hello,

Could you elaborate a little bit? Would you like to encode a model as a statechart in order to test it, or to use a statechart expressing the tests that your model should pass?

Usually a Django model represents the data where a statechart represents the process so I'm not convinced using the latter to encode/test the former is really appropriate.

**Comment by simkimsia** [comment]
Thank you, @AlexandreDecan for the quick response.

> Could you elaborate a little bit? Would you like to encode a model as a statechart in order to test it, or to use a statechart expressing the tests that your model should pass?


I have an actual usecase where I have two Django models: `Order` and `OrderLineItem`

- One `Order` has zero or many `OrderLineItem`s.
- One `OrderLineItem` MUST belong to a maximum of one `Order`.
- Each `Order` has a `status` field containing values such as : `DRAFT`, `PENDING_PO`, `PROCESSING`, `PENDING_GR`, `GR`, `CANCELLED`, `DELETED`.

I left out more details but I want to keep my example simple for now.

So the `Order` will be transitioning between multiple states, represented by the `status` field.


I like to represent the behavior of the process, as you put it, when the `Order` goes through different states depending on the preconditions and postconditions.

I like to represent the state transitions (either as yaml or python API) as a way to write BDD tests on **how a Django model with state transitions** is supposed to behave?

Am I making any sense?

**Comment by simkimsia** [comment]
Separately, but I can assure you it's related to the question, 

when I look at the files for the elevator example

https://github.com/AlexandreDecan/sismic/tree/master/docs/examples/elevator

I saw .graffle, .png, .feature, .yaml, etc

I want to know if I were to use sismic for my Django example, which file do I generate first? And in which sequence do I then generate the rest as "derivatives" of the first file?

I also don't mind redoing the elevator example as a Django model so that it's easier to grok sismic as a Django developer.

If I can redo the elevator example (or my `Order` hypothetical example) as a Django model, then the separate questions about which file to generate first, will also be answered by the exercise.

**Comment by AlexandreDecan** [comment]
What about using a dedicated Django extension for that? (see e.g. https://pypi.org/project/django-states/)? 

You can "easily" encode your model within a Sismic statechart and see whether the various triggers adequately change the state. The problem being to map "internal events" from Django to events in Sismic and vice-versa. To summarize, I don't really see how to integrate a (sismic) statechart into a django model or vice-versa, especially for testing django models since they rely on a lot of internalson which we have little to no control. 

Having said that, and based on your last message, if you are willing to rewrite your Django models as Sismic statecharts, it should be possible to take advantage of Sismic's testing abilities... but you may run into a lot of problems because Django models are a bit (a lot) "magical" in the way they work, and I don't know how easy it will be to get around this magic to have a "real statechart" while keeping some of this "magic" to comply with Django internals :-)

**Comment by AlexandreDecan** [comment]
> I want to know if I were to use sismic for my Django example, which file do I generate first? And in which sequence do I then generate the rest as "derivatives" of the first file?

The first file to generate is the yaml file encoding the statechart, then some boilerplate to integrate Sismic into a Django model, then the various files to test the statechart (i.e., BDD scenarios as .feature, property statecharts, etc. depending on your needs). 

But again, I'm not really convinced it's a good idea ;-)

**Comment by simkimsia** [comment]
> if you are willing to rewrite your Django models as Sismic statecharts,

Oh, seems like I misunderstood something about sismic.

Let me change the question to another angle.

> The first file to generate is the yaml file encoding the statechart, then some boilerplate to integrate Sismic into a Django model, then the various files to test the statechart (i.e., BDD scenarios as .feature, property statecharts, etc. depending on your needs).

So for your BDD example, you generate the yaml AND the .feature by hand. Then you use the .feature to test whether the .yaml fulfils the tests as written in the .feature.

Am I right?

**Comment by simkimsia** [comment]
Sorry @AlexandreDecan that I now digress about the stopwatch example

In the stopwatch example, you have the following files

<img width="575" alt="image" src="https://github.com/AlexandreDecan/sismic/assets/245021/b9389eef-ee88-421e-b4a8-5db0d0fbfcdb">

Unlike your BDD example, there are now .py files. Which I assume are implementation files to be tested.

Am I right to say that:
- the stopwatch.yaml is written by hand
- the stopwatch.py is also written by hand
- the purpose of stopwatch.yaml is to test that the stopwatch.py meets the requirements?

Or the .yaml file will generate the .py files as derivatives?

I am also looking at the differences between stopwatch.yaml and stopwatch_external.yaml

<img width="1137" alt="image" src="https://github.com/AlexandreDecan/sismic/assets/245021/52223035-1f23-475e-9593-26321f08e5ee">


Clearly, the stopwatch_external.yaml expects a "stopwatch" object. Unlike the stopwatch.yaml.

So, is it more correct to say that

1. stopwatch.yaml generates stopwatch.py
2. and then stopwatch_external.yaml then tests stopwatch.py 

?

Thank you.

**Comment by AlexandreDecan** [comment]
Basically, all you need to have a statechart is a yaml file describing the behaviour of this statechart. In many cases though, you want the statechart to interact with other components (e.g., a database, a UI, etc.). You have two options: either you implement this external component as part of your statechart (or you pass it to the statechart Interpreter), that's what we did for "stopwatch_external.yaml" that receives the component and manipulate it. The other option is to rely on the events sent and received by the statechart to interact with it. That's what we did for "stopwatch.yaml", orchestrated from "stopwatch.py". 

As soon as you have a working statechart (regardless of the above), you can define tests for this statechart in multiple ways: 
 - By writing "traditional" unit tests in Python, using the Sismic API; 
 - By specifying contracts within the statechart: these contracts will be checked every time the statechart is executed; 
 - By specifying property statecharts (statecharts following some conventions) and binding them to the original statechart so that these properties are checked at runtime (similarly to what is done with contracts); 
 - By specifying "feature files" that describe the expected behaviour of the statechart, and by executing these feature files with Sismic. Notice that these feature files are mostly "unit tests written in English": they express different scenarii and the expected output/state for each of them. 

TL;DR: For your sentence "*In other words, the .yaml is the implementation of the statechart. The .feature is the test on whether the statechart is correctly implemented.*", the answer is **yes** :-)

**Comment by simkimsia** [comment]
Thank you for your detailed explanation and your patience, @AlexandreDecan 

Just want to say upfront, this thread to me is async, so don't feel any pressure to respond to me quickly.

I got 1 broad question and a series of more in-depth questions regarding the stopwatch example

### Broad Question

This means that sismic has 3 ways to implement statecharts and 4 ways to test/check the statecharts. 

3 ways to implement:
- embed the code directly into the yaml (like the elevator example)
- rely on external component by pass in via initial context e.g. stopwatch_external)
- rely on events to and from statechart (e.g. the stopwatch.yaml example)

4 ways to check/test:
- traditional unit tests in Python using SISMIC API, so only check during executing the tests
- BDD feature files, so only check during executing the tests
- contracts within statecharts, so check during runtime every time
- property statecharts that are bound to the original process statecharts, so check during runtime every time

Yes?

So theoretically, there are 12 ways to use SISMIC (assuming both check and implement are done with help of SISMIC and ignoring hybrids)

### In-depth, Narrower Questions

All to do with the stopwatch example

With the help of your explanation and re-reading the docs that cover stopwatch example at https://sismic.readthedocs.io/en/latest/integration.html

what I get is

1. control statechart from environment (aka rely on events  to and from statechart)
  - stopwatch_gui.py (this is the environment)
  - stopwatch.yaml (this is the statechart)
  
2. control environment from inside statechart (aka rely on external component by pass in via initial context )
  - stopwatch_external_gui.py (this is environment)
  - stopwatch_external.yaml (this is the statechart)
  - stopwatch.py (this is the external component passed into the initial_context)

Did I summarize the above correctly?

What about the graffle?

I surmise the .png files are derived from the graffle. Can the SISMIC help to generate the graffle from the .yaml statecharts or you have to hand generate that using a different software?

**Comment by AlexandreDecan** [comment]
> This means that sismic has 3 ways to implement statecharts and 4 ways to test/check the statecharts.
> 3 ways to implement:
>     * embed the code directly into the yaml (like the elevator example)
>     * rely on external component by pass in via initial context e.g. stopwatch_external)
>     * rely on events to and from statechart (e.g. the stopwatch.yaml example)

Strictly speaking, there are only two ways to write a statechart: 
 - Using a yaml description of the statechart; 
 - Using the model API.

The former is the preferable way, the latter is mostly for internal purposes. 

The three items you describe do not correspond to "ways to write a statechart", but ways to make statecharts and other components communicate. 

> 4 ways to check/test:
> Yes?

Yes :-)
 
> what I get is
>     1. control statechart from environment (aka rely on events  to and from statechart)
>     * stopwatch_gui.py (this is the environment)
>     * stopwatch.yaml (this is the statechart)

Assuming a MVC model, the M+C is written within the statechart, and V is the gui (in Python).

>     2. control environment from inside statechart (aka rely on external component by pass in via initial context )
>     * stopwatch_external_gui.py (this is environment)
>     * stopwatch_external.yaml (this is the statechart)
>     * stopwatch.py (this is the external component passed into the initial_context)

In this case, the V is still the gui (in Python), the M is stopwatch.py, and the C is the statechart in yaml. 

> What about the graffle?

It wasn't generated by Sismic, but by another tool, manually. 

> I wonder if the graffle is generated via some of these extensions?

No, but we support PlantUML directly in Sismic. See the documentation here:
https://sismic.readthedocs.io/en/latest/visualization.html

**Comment by AlexandreDecan** [comment]
Can I close this issue?

**Comment by simkimsia** [comment]
Yes, thank you, @AlexandreDecan  🙇🏻‍♂️

**Comment by AlexandreDecan** [comment]
You're welcome ;-) Feel free to open another issue if you have questions ;-)
